Simon-Initiative/oli-torus
# Grade services

This PR adds support for LMS grade passback via LTI Assignment and Grade Services (LTI AGS) 2.0. 

## The approach

1. A student submits a graded page attempt.
2. The system finalizes that attempt and rolls up a score to the `resource_access` record for that resource and student.
3. If the LTI Launch params have LTI AGS enabled, the system requests from the LMS an access token.
4. Using the access token as a Bearer token, the system queries the AGS endpoint to determine if a line item for this resource exists.  
5. If the line item does not exist, the system uses the access token to invoke another AGS endpoint to create that line item.
6. If the line item does exist, but it's label or max score relative to the current score is out of date, the system will issue a request to an AGS endpoint to update the line item. 
7. Finally, the system uses the access token and the /scores AGS endpoint to send the score for this student to this line item

## Implications

1. Line items will not appear in the LMS gradebook until the first student completes one attempt.  
2. Publications that change graded page titles will not be reflected in the grade book until a student completes an attempt.  
3. The line item max score will be updated to reflect the state of the max of the most recently posted score - for any student. But this could be problematic if an instructor makes a change to an assessment to add a new question, after one student has already finalized an attempt with the original number of questions. 

The first two from above are okay for now, but we may want to add a feature to allow an instructor to do a full update of line items in the gradebook, at any time.  The third one requires more investigation to understand what LMSes do with the line item max score, as compared to individual student scores which may have a different max.  I am going to do this investigation over the next day and then possibly issue a follow up PR to adjust this score maximum handling. 

Closes #594 

## Testing This

1. Create an LTI developer key in Canvas as usual.
2. Enable AGS access (the first five or so permissions under the LTI Advantage options) in the Developer Key in Canvas
3. Update your `oli.env` environment variables to set your `HOST` to be the full protocol and host of your ngrok served local Torus instance. For example: `HOST=https://c51bd1ea5000.ngrok.io`.  Restart your Dev environment.
4. Publish a course with a graded page.
5. Launch into it from Canvas, and take the graded assignment.  You should immediately see the grade appear back in the Canvas gradebook
